[SPARK-10474] [SQL] Aggregation fails to allocate memory for pointer array (round 2)

andrewor14 opened the pull request:
This patch reverts most of the changes in a previous fix #8827.

The real cause of the issue is that in `TungstenAggregate`'s prepare method we only reserve 1 page, but later when we switch to sort-based aggregation we try to acquire 1 page AND a pointer array. The longer-term fix should be to reserve also the pointer array, but for now **_we will simply not track the pointer array**_. (Note that elsewhere we already don't track the pointer array, e.g. [here](https://github.com/apache/spark/blob/a18208047f06a4244703c17023bb20cbe1f59d73/sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java#L88))

Note: This patch reuses the unit test added in #8827 so it doesn't show up in the diff.

Commit:
Revert "[SPARK-10474] [SQL] Aggregation fails to allocate memory for pointer array"

This reverts commit 7ff8d68cc19299e16dedfd819b9e96480fa6cf44.

Commit:
Add back test

Commit:
Do not track pointer array...

Commit:
Do not release pointer array memory since we don't track it

Commit:
Clarify comment

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Use correct JIRA number in comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42924 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42924/consoleFull) for   PR 8888 at commit [`c910d0b`](https://github.com/apache/spark/commit/c910d0b46d54b549babd6b72c8ed032c61bc331d).

davies commented:
LGTM

Commit:
Correct a comment (minor)

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42922/
Test FAILed.

SparkQA commented:
[Test build #1799 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1799/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #42925 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42925/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #1800 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1800/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #1801 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1801/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #1802 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1802/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #42924 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42924/console) for   PR 8888 at commit [`c910d0b`](https://github.com/apache/spark/commit/c910d0b46d54b549babd6b72c8ed032c61bc331d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42924/
Test FAILed.

SparkQA commented:
[Test build #42925 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42925/console) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42925/
Test FAILed.

Commit:
Fix tests

Looks like we still tracked the pointer array memory when we grow
it. Don't do that.

SparkQA commented:
[Test build #1803 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1803/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #1804 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1804/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #1799 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1799/console) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1806 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1806/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #42942 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42942/consoleFull) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).

SparkQA commented:
[Test build #1805 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1805/consoleFull) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).

SparkQA commented:
[Test build #1800 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1800/console) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1802 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1802/console) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1801 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1801/console) for   PR 8888 at commit [`ed36351`](https://github.com/apache/spark/commit/ed36351dfbe7bc3cf88693aa784131319063089f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

andrewor14 commented:
By the way @JoshRosen this seems to run core tests like `SecurityManagerSuite` even though it touches only SQL stuff. Just FYI.

chenghao-intel commented:
https://github.com/apache/spark/blob/master/core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java#L138 Seems reserve the data page only if the `existingInMemorySorter` is unset, however, SPARK-474 is not this case right? As it follows the sort merge join operator.

JoshRosen commented:
@andrewor14, it's running core tests because you changed `core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java`

andrewor14 commented:
@chenghao-intel SPARK-10474 is caused by an aggregate falling back to sort-based aggregation. In this case we don't acquire the page in the constructor, but we do acquire it when we insert into the sorter later.

SparkQA commented:
[Test build #1808 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1808/consoleFull) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).

SparkQA commented:
[Test build #1810 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1810/consoleFull) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).

andrewor14 commented:
>  it's running core tests because you changed core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

Ah yes never mind.

SparkQA commented:
[Test build #1807 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1807/consoleFull) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).

SparkQA commented:
[Test build #1811 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1811/consoleFull) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).

SparkQA commented:
[Test build #1809 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1809/consoleFull) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).

chenghao-intel commented:
@andrewor14 that's actually what I mean, if we didn't reserve the memory when creating the down streaming operator, we probably never get the chance to acquire the page when inserting records to `sorter`, as the upstreaming operator (like SMJ) probably eat out all of the memory, which even lead to the hash aggregation switching to sort-based aggregation when first record comes. Do you think that's possible?

andrewor14 commented:
@chenghao-intel but we do reserve the page in advance. See `TungstenAggregate`: https://github.com/apache/spark/blob/084e4e126211d74a79e8dbd2d0e604dd3c650822/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala#L77

SparkQA commented:
[Test build #42942 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42942/console) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42942/
Test PASSed.

andrewor14 commented:
OK it passed tests. I'm merging this into master 1.5.

SparkQA commented:
[Test build #1808 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1808/console) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1811 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1811/console) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1810 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1810/console) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1809 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1809/console) for   PR 8888 at commit [`2dc34c3`](https://github.com/apache/spark/commit/2dc34c3b8ed73b3a0d223e2cd0fc26b740f442d5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

chenghao-intel commented:
Thank you @andrewor14 for the explanation. I believe you're talking about to reserve the data page in advanced via the `UnsafeFixedWidthAggregationMap.map` (BytesToBytesMap), which will reserve the data page in its constructor. 
Everything looks reasonable to me now, the only concern is the performance, since only a single data page available for the sort-based aggregation. Anyway, it's not related to this fix.

andrewor14 commented:
Yes, it's not related. What you mention here is a bigger problem. The current solution only ensures that we don't starve any operators. In the future we can improve this mechanism by introducing some force spilling mechanism, but that's too big of a change to backport to 1.5.
